mlir/lib/Dialect/MIOpen/Transforms/ThreadwiseGemmLowering.cpp

```diff
@@ -212,119 +212,123 @@ struct XdlopsGemmV2RewritePattern : public OpConversionPattern<XdlopsGemmV2Op> {
-    // for(index_t k_i = 0; k_i < KPerThread; ++k_i) {
-    //   matrixAElement = a[k_i];
-    //   matrixBElement = b[k_i];
-    //   // Loop within a kpack
-    //   for(index_t ki_i = 0; ki_i < k_base * KRepeats; ki_i += k_base)
-    //     argA = &matrixAElement[ki_i];
-    //     argB = &matrixAElement[ki_i];
-    //     p_c_thread = mfma_type.template run<MPerXlops * MRepeats,
-    //                                         NPerXdlops * NRepeats,
-    //                                         AStride,
-    //                                         BStride>(argA, argB,
-    //       p_c_thread);
-    // }
-    int64_t outerLoopUpperBound = KPerThread;
-    // In case xdlops consume multiple elements from memref<T>, divide by k_base
-    // such that the generated loop is not out of bounds.
-    if (KPack == 1) {
-      outerLoopUpperBound /= k_base;
-    }
     auto regAConstantOp =
         b.create<ConstantIndexOp>(loc, op.regOffsetAAttr().getInt());
     auto regBConstantOp =
         b.create<ConstantIndexOp>(loc, op.regOffsetBAttr().getInt());
 
-    auto outerLoop = b.create<AffineForOp>(loc, 0, outerLoopUpperBound);
-    auto outerLoopb = ConversionPatternRewriter::atBlockBegin(
-        outerLoop.getBody(), b.getListener());
-    auto outerLoopiv = outerLoop.getInductionVar();
+    auto populateMfma = [&](OpBuilder &b, Value &argA, Value &argB) {
+      for (int64_t i = 0; i < vectorNumber; ++i) {
+        // Note below is assuming only one of MRepeats or NRepeats is larger
+        // than 1, which fits the existing blockwisegemmv2op implementation.
+        // TODO: Move MRepeats and NRepeats into xdlopsgemmv2op
+        int64_t regDOffset = 0;
+        if (op.regOffsetAAttr().getInt() > 0 ||
+            op.regOffsetBAttr().getInt() > 0) {
+          regDOffset += vectorNumber;
+        }
+        Value offset =
+            b.createOrFold<arith::ConstantIndexOp>(loc, regDOffset + i);
+
+        auto vectorC = b.create<memref::LoadOp>(loc, vectorType,
+                                                adaptor.matrixC(), offset);
+        auto mfma = b.create<amdgpu::MFMAOp>(
+            loc, vectorType, mfmaInstr, argA, argB, vectorC,
+            /*cbsz=*/imms[i][0], /*abid=*/imms[i][1], /*blgp=*/imms[i][2]);
+        auto vectorD = mfma.destD();
+
+        b.create<memref::StoreOp>(loc, vectorD, adaptor.matrixC(), offset);
+      }
+    };
 
-    // matrixAElement and matrixBElement are only useful when KPack > 1
-    Value matrixAElement;
-    Value matrixBElement;
+    // TODO: zyin adopt generic layout: K, KRepeats, k_base
+    // After that, we'd be able to uniform between kPack/nonKPack
+    if (KPack == 1) {
+      // for(index_t k_i = 0; k_i < KPerThread; k_i += k_base) {
+      //   argA = a[k_i];
+      //   argB = a[k_i];
+      //   p_c_thread = mfma(argA, argB, p_c_thread);
+      // }
+      Value zeroConstantOp = b.create<ConstantIndexOp>(loc, 0);
```

Comment:
Maybe a `createOrFold` to save the pattern rewriter some work?

Reply:
Sounds good.

Reply:
Done in 1405899